classy-afrovenator: Add billing tab to /settings

taravancil opened the pull request:
## Links
* [classy-afrovenator.glitch.me](https://classy-afrovenator.glitch.me)
* [[ch6209]](https://app.clubhouse.io/glitch/story/6209/add-billing-tab-to-settings): Add billing tab to /settings

## GIF/Screenshots:
**To come. The backend is returning `500`s for the `/glitchPro` endpoints. Will update when Justin is in and can investigate.**

## Changes:
This adds a "Subscription" tab to /settings behind the `pufferfish` feature flag.

## How To Test:
*Note that this is the dogfood implementation of this feature, so review should be focused on "Does it work well enough to test internally and iterate on?" rather than polish and cleverness.*

### Testing Subscription Tab Visibility
To see the tab:

1. Log in
2. Make sure you're in [@testing-team](https://glitch.com/@testing-team)
3. Make sure the `pufferfish` flag is enabled on [classy-afrovenator.glitch.me/secret](https://classy-afrovenator.glitch.me/secret)

| Case  | Expected Outcome | 
| ------------- | ------------- |
| In Testing Team + Enabled the feature on /secret  | Subscription tab visible |
| In Testing Team + Disabled the feature on /secret  | Subscription tab NOT visible |
| NOT In Testing Team + Enabled the feature on /secret  | Subscription NOT tab visible  |
| NOT In Testing Team + Disabled the feature on /secret  | Subscription NOT tab visible |

### Testing Creating + Cancelling a Subscription
**To come. The backend is returning `500`s for the `/glitchPro` endpoints. Will update when Justin is in and can investigate.**

## Feedback I'm looking for:
- Can you create a subscription with test data?
- Can you cancel that subscription?
- Do the `useStripe` and `useSubscription` hooks feel ok to you?

taravancil commented on src/state/stripe.js:
```diff
@@ -0,0 +1,15 @@
+  const [loaded] = useScript(stripeJS);
+  useEffect(() => {
+    if (loaded) {
+      setStripe(window.Stripe('pk_test_WIJxKl0T1xT8zSb2StHu8zlo'));
```
Just noting that this is a publishable key. 😄

taravancil commented:
/cc @FogCreek/pufferfish

sarahzinger commented on src/components/account-settings-overlay/subscription-settings.js:
```diff
@@ -0,0 +1,62 @@
+
+  async function subscribe() {
+    try {
+      const { data } = await createSubscriptionSession({ successUrl: 'https://glitch.com', cancelUrl: 'https://glitch.com/pricing' });
```
apologies if you're already on it, but just in case I think these urls don't feel quite right, yeah? Not 100% sure what we should do for dogfood, we could keep em on the pricing page with some kind of flash of success? Or redirect them back to their user profile so they see the new collection?

taravancil replied:
Yeah, I just put those as placeholders until we have proper mockups and stuff. Does that feel ok to you?

sarahzinger replied:
yeah that's fine! Can I vote for now we go to the user page if it succeeds and stay on the billing page if it fails?

sarahzinger commented on src/hooks/use-script.js:
```diff
@@ -0,0 +1,34 @@
+    const script = document.createElement('script');
+
+    if (loadedScripts.has(src)) {
+      setLoaded(true);
```
could you not have a script tag with a src that wasn't loaded? like it errored or something? Then again this is probably 💯 for dogfood imo!

taravancil replied:
Yes 😝 In fact if you simulate a slow network you will probably see how this fails. But for dogfood I think it's fine.

sarahzinger replied:
cool maybe we should just leave a comment so we come back to it later

sarahzinger commented on src/presenters/pages/settings.js:
```diff
@@ -15,35 +18,75 @@ import styles from './settings.styl';
+              <Tab key="subscription" hidden={!showSubscriptionTab} className={styles.settingsTab}>
+                <Button as="span">Subscription</Button>
+              </Tab>
+              {/* <Button disabled onClick>Privacy & Notifications</Button> */}
```
I think we can kill this comment now btw

whimsicallyson commented on src/components/account-settings-overlay/subscription-settings.js:
```diff
@@ -0,0 +1,62 @@
+        </>
+      ) : (
+        <>
+          <Text defaultMargin>Subscribe to the Extra Memory monthly plan for $14 per month.</Text>
```
when I click through to Stripe checkout, it says $10/mo?

taravancil replied:
This is all placeholder content for the dogfood stage :)

whimsicallyson replied:
fair!

whimsicallyson commented on src/presenters/pages/settings.js:
```diff
@@ -15,35 +18,74 @@ import styles from './settings.styl';
+            <TabList className={styles.settingsActions}>
+              <Tab key="account" className={styles.settingsTab}>
+                <Button as="span">Account</Button>
+              </Tab>
```
I want more whitespace between these buttons, but I get that this is dogfooding and that's not a priority. just stating for the record :)

taravancil replied:
Yep 😄 We don't have proper mockups yet, so everything is a bit ugly and more about testing Stripe and such.

taravancil commented:
@sarahzinger thanks! I'm working on tests, but will let you know when it's ready.

taravancil commented:
@sarahzinger this is ready for another round of review!

Changed since your last review:

- Refactored the settings page to use a `<SettingsTabsContainer>` and `<SettingsTabs>` component, for the sake of making it possible to test the tabs by passing in props
- Updated the Stripe `successUrl` and  `cancelUrl`
- Added a note about the `useScript` edge case

sarahzinger commented on test/components/settings-tabs.js:
```diff
@@ -0,0 +1,76 @@
+          .find(SettingsTabs)
+          .find('li[role="tab"]')
+          .at(1)
+          .prop('hidden'),
```
I think if we ever update the tab order this test will no longer be testing the right thing yeah? Would it make sense to add a data attribute to make it easier to test the tabs?
